aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,602 @@
+            error_index,
+            " (output volumes are of size ",
+            oT,
+            "x",
+            oH,
+            "x",
+            oW);
+      }
+    }
+  }
+  return output;
+}
+
+void max_unpooling3d_shape_check(
+    const Tensor& input,
+    const Tensor& gradOutput,
+    const Tensor& indices,
+    IntList output_size,
+    IntList stride,
+    IntList padding) {
+  int64_t oT = output_size[0];
+  int64_t oW = output_size[1];
+  int64_t oH = output_size[2];
+  AT_CHECK(
+      indices.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type Long");
+  AT_CHECK(
+      (input.ndimension() == 4 || input.ndimension() == 5),
+      "Input to max_unpooling3d should be a 4d or 5d Tensor",
+      input.sizes());
+  AT_CHECK(
+      output_size.size() == 3,
+      "There should be exactly three elements (depth, width, height) in output_size");
+  AT_CHECK(
+      stride.size() == 3,
+      "There should be exactly three elements (depth, width, height) in stride");
+  AT_CHECK(
+      padding.size() == 3,
+      "There should be exactly three elements (depth, width, height) in padding");
+  AT_CHECK(
+      input.sizes() == indices.sizes(),
+      "Shape of indices should match shape of input");
+
+  AT_CHECK(input.numel() > 0, "Input must be non-empty");
+
+  AT_CHECK(
+      stride[0] > 0 && stride[1] > 0 && stride[2] > 0,
+      "stride should be never greater than zero, but got stride: ",
+      stride);
+
+  int dimw = 3;
+  int dimh = 2;
+  int dimt = 1;
+  int dimn = 0;
+
+  if (input.ndimension() == 5) {
+    dimw++;
+    dimh++;
+    dimt++;
+    dimn++;
+  }
+
+  int nslices = input.size(dimn);
+
+  if (gradOutput.defined()) {
+    if (oT != gradOutput.size(dimt) ||
+        oH != gradOutput.size(dimh) ||
+        oW != gradOutput.size(dimw)) {
+      AT_ERROR(
+          "Inconsistent gradOutput size. oT= ",
+          oT,
+          ", oH= ",
+          oH,
+          ", oW= ",
+          oW,
+          ". gradOutput: ",
+          gradOutput.size(dimt),
+          "x",
+          gradOutput.size(dimh),
+          "x",
+          gradOutput.size(dimw));
+    }
+    AT_CHECK(
+        gradOutput.ndimension() == input.ndimension() &&
+            gradOutput.size(dimn) == nslices,
+        "gradOutput and input Tensors should have same number of dimensions and also the same number of channels/slices");
+  }
+}
+
+Tensor& max_unpooling3d_forward_out_cpu(
+    Tensor& output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size,
+    IntList stride,
+    IntList padding) {
+  int64_t oT = output_size[0];
+  int64_t oW = output_size[1];
+  int64_t oH = output_size[2];
+  max_unpooling3d_shape_check(
+      self, Tensor(), indices, output_size, stride, padding);
+
+  if (self.ndimension() == 5) {
+    output.resize_({self.size(0),
+                    self.size(1),
+                    oT,
+                    oH,
+                    oW});
```

Comment:
There is probably a legit bug in the original implementation here. The passed in output sizes are in the order TxWxH. but then we resize the output as TxHxW. This appears to be exactly what the old code did, strangely enough. @andersenchen do you mind filing a bug for this?

Reply:
Did we get a bug report for this?

Reply:
@andersenchen Actually, this is not a bug, because our code generator apparently reorders arguments when calling these kernels. TxHxW is the correct ordering, and the argument order is OK too. Phew!